### PR TITLE
Hyperparameter Optimization Module

### PR DESCRIPTION
**Pull request by middleyuan**
The pull request contains 
1. DDPG, SAC, PPO

- Change the saving method for the checkpoint.
- Add activation as an argument for the constructor.

2. GPMPC

- Add the feature that allows one to learn GP in parallel (this speeds up roughly D times faster, where D is the dimension of the dynamics)
- Add activation as an argument for the constructor ('Matern' or 'RBF' kernel are supported).

3. Hyperparameter optimization (HPO) module

- database.py implements the functions that interact with MySQL.
- hpo_sampler.py defines the hyperparameter spaces for the selected controller.
- hpo.py defines the hyperparameter optimization class.

4. safe_control_gym/hyperparameters/README.md describes the basic steps for running HPO.

### CODE REVIEW

**Comment by adamhall** [comment]
Run precommit hooks etc. to ensure formatting is good.

**Comment by adamhall** [comment]
`experiments` folder was removed. Clean up experiments and move to `examples` folder that exists in the current main branch.

**Comment by adamhall** [comment]
Move any files used for `tests` into the `tests` directory.

**Comment by adamhall** [comment]
rename `hyperparameters_13.....yaml` to something more readable.

File: examples/hpo/gp_mpc/config_overrides/cartpole/gp_mpc_cartpole_hpo_.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+hpo_config:
+
+  hpo: True # do hyperparameter optimization
+  load_if_exists: True # this should set to True if hpo is run in parallel
```
Comment (Federico-PizarroBejarano) [inline examples/hpo/gp_mpc/config_overrides/cartpole/gp_mpc_cartpole_hpo_.yaml:4]: `load_if_exists` is a confusing name, what does it mean?

File: examples/hpo/gp_mpc/config_overrides/cartpole/gp_mpc_cartpole_hpo_.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  use_database: False # this is set to true if MySQL is used
+  objective: [exponentiated_avg_return] # [other metrics defined in base_experiment.py]
```
Comment (Federico-PizarroBejarano) [inline examples/hpo/gp_mpc/config_overrides/cartpole/gp_mpc_cartpole_hpo_.yaml:6]: why are `objective` and `direction` lists?

File: examples/hpo/gp_mpc/config_overrides/cartpole/cartpole_stab.yaml [hunk 3]
```diff
@@ -0,0 +1,67 @@
+task_config:
```
Comment (Federico-PizarroBejarano) [inline examples/hpo/gp_mpc/config_overrides/cartpole/cartpole_stab.yaml:1]: Why is the file structure of this example `algo/config_override/env`? All of these files are configs rt, so the directory structure should be `config_overrides/algo/env` or something like that, with `config_override` at the top level.

File: examples/hpo/hpo_experiment.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    results, metrics = experiment.run_evaluation(n_episodes=1, n_steps=None, done_on_max_steps=True)
+    control_agent.close()
+
+    if True:
+        if config.task == Environment.CARTPOLE:
+            graph1_1 = 2
+            graph1_2 = 3
+            graph3_1 = 0
+            graph3_2 = 1
+        elif config.task == 'quadrotor_2D':
```
Comment (Federico-PizarroBejarano) [inline examples/hpo/hpo_experiment.py:104]: You haven't added quadrotor examples yet rt? I would remove this extra code until its necessar

File: safe_control_gym/controllers/ppo/ppo_utils.py [hunk 5]
```diff
@@ -117,6 +117,8 @@ def update(self,
         '''Updates model parameters based on current training batch.'''
         results = defaultdict(list)
         num_mini_batch = rollouts.max_length * rollouts.batch_size // self.mini_batch_size
+        # assert if num_mini_batch is 0
```
Comment (Federico-PizarroBejarano) [inline safe_control_gym/controllers/ppo/ppo_utils.py:120]: should be "asser num_mini_batches is not 0" right?

File: safe_control_gym/experiments/base_experiment.py [hunk 6]
```diff
@@ -405,7 +413,7 @@ def compute_metrics(self, data, verbose=False):
         }
         return metrics
 
-    def get_episode_data(self, key, postprocess_func=lambda x: x):
+    def get_episode_data(self, key, postprocess_func=lambda x: x, exponentiate=False):
```
Comment (Federico-PizarroBejarano) [inline safe_control_gym/experiments/base_experiment.py:416]: Should add this `exponentiate` argument to the Args list. But more importantly, what is this and why is it needed?

File: safe_control_gym/experiments/base_experiment.py [hunk 7]
```diff
@@ -441,7 +455,15 @@ def get_episode_lengths(self):
...
-    def get_episode_returns(self):
+    def get_episode_returns(self, exponentiate=False):
+        '''Total reward/return of episodes.
+
+        Returns:
+            episode_rewards (list): The total reward of each episode.
+        '''
+        return self.get_episode_data('reward', postprocess_func=sum, exponentiate=exponentiate)
+    
+    def get_episode_exponentiated_returns(self):
```
Comment (Federico-PizarroBejarano) [inline safe_control_gym/experiments/base_experiment.py:466]: this function is called exponentiated but it is not exponentiated

File: safe_control_gym/experiments/base_experiment.py (hunk 6 above)
Comment (Federico-PizarroBejarano) [reply safe_control_gym/experiments/base_experiment.py:416]: If it is only for returns, why not just create a different postprocess_func?

File: safe_control_gym/hyperparameters/README.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+[`README.md`](README.md)
+
+
+To perform hyperparmeter optimization, you may need `MySQL` database (optional):
+
+```bash
+sudo apt-get install mysql-server # MySQL server starts automatically after the installation
+```
+
+To set up, run the following commands sequencially:
```
Comment (Federico-PizarroBejarano) [inline safe_control_gym/hyperparameters/README.md:17]: sequentially

File: safe_control_gym/hyperparameters/README.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+```bash
+pip install mysql-connector-python
+pip install pymysql
+```
+
+For HPO, `optuna` is used, install the following packages:
+
+```bash
+pip install optuna
```
Comment (Federico-PizarroBejarano) [inline safe_control_gym/hyperparameters/README.md:37]: shouldnt be necessary considering your pyproject.toml changes rt

File: safe_control_gym/hyperparameters/README.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+                    --overrides, \
+                    ./examples/hpo/rl/config_overrides/{sys}/{sys}_{task}.yaml, \
+                    ./examples/hpo/rl/{algo}/config_overrides/{sys}/{algo}_{sys}_.yaml, \
+                    ./examples/hpo/rl/{algo}/config_overrides/{sys}/{algo}_{sys}_hpo_.yaml, \
+                    --output_dir, {output_dir}, \
+                    --seed, 7, \
+                    --use_gpu, True
+```
+
+You may replace `{sys}` with `carpole` in order to run the script.
```
Comment (Federico-PizarroBejarano) [inline safe_control_gym/hyperparameters/README.md:63]: cartpole

File: safe_control_gym/hyperparameters/README.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+                    --overrides \
+                    ./examples/hpo/rl/{algo}/config_overrides/cartpole/{algo}_{sys}_.yaml \
+                    ./examples/hpo/rl/config_overrides/{sys}/{sys}_{task}.yaml \
+                    --output_dir {output_dir} \
+                    --tag {run_name} \
+                    --opt_hps {best_hp_file} \
+                    --task {sys} --seed 2 --use_gpu True
+```
+
+You may replace `{sys}` with `carpole` in order to run the script. As an example, `{best_hp_file}` can be replaced with `examples/hpo/rl/sac/config_overrides/cartpole/optimized_hyperparameters.yaml` for `sac`.
```
Comment (Federico-PizarroBejarano) [inline safe_control_gym/hyperparameters/README.md:81]: cartpole

File: safe_control_gym/hyperparameters/hpo.py [hunk 12]
```diff
@@ -0,0 +1,306 @@
+""" The implementation of HPO class
+
+Reference:
+    * stable baselines3 https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/rl_zoo3/hyperparams_opt.py
+    * Optuna: https://optuna.org
+
+Reruirement:
```
Comment (Federico-PizarroBejarano) [inline safe_control_gym/hyperparameters/hpo.py:7]: Requirement. Also, this is installed by the pyproject.toml rt so no need to mention

**Review by Federico-PizarroBejarano — COMMENTED** [review]
I pulled your branch and reran the pre-commit hooks and there are some minor fixes you should add.

File: .gitignore [hunk 13]
```diff
@@ -9,7 +9,12 @@ examples/pid/*data/
 results/
 z_docstring.py
 TODOs.md
-
+#
+hpo_strategy_study*/
+*.jpg
```
[the remark this answers is not in the document]
Comment (middleyuan) [reply .gitignore]: Should be already removed?

File: .pre-commit-config.yaml [hunk 14]
```diff
@@ -0,0 +1,58 @@
+ # Install the pre-commit hooks below with
```
[the remark this answers is not in the document]
Comment (middleyuan) [reply .pre-commit-config.yaml]: This file doesn't show when I look at git diff. No idea what the issue is?

File: examples/no_controller/verbose_api.py [hunk 15]
```diff
@@ -5,8 +5,9 @@
     $ python3 verbose_api.py --task quadrotor --overrides verbose_api.yaml
 '''
 
-import time
 import inspect
```
[the remark this answers is not in the document]
Comment (middleyuan) [reply examples/no_controller/verbose_api.py]: For me, it also doesn't show up when I look at git diff.

File: examples/hpo/gp_mpc/config_overrides/cartpole/gp_mpc_cartpole_150.yaml [hunk 16]
```diff
@@ -0,0 +1,66 @@
+algo: gp_mpc
```
[the remark this answers is not in the document]
Comment (Federico-PizarroBejarano) [reply examples/hpo/gp_mpc/config_overrides/cartpole/gp_mpc_cartpole_150.yaml:1]: Sounds good

File: examples/no_controller/verbose_api.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Federico-PizarroBejarano) [reply examples/no_controller/verbose_api.py]: Weird

File: examples/hpo/gp_mpc/config_overrides/cartpole/gp_mpc_cartpole_hpo_.yaml (hunk 1 above)
Comment (Federico-PizarroBejarano) [reply examples/hpo/gp_mpc/config_overrides/cartpole/gp_mpc_cartpole_hpo_.yaml]: Ah ok sounds good

File: examples/hpo/gp_mpc/config_overrides/cartpole/gp_mpc_cartpole_hpo_.yaml (hunk 2 above)
Comment (Federico-PizarroBejarano) [reply examples/hpo/gp_mpc/config_overrides/cartpole/gp_mpc_cartpole_hpo_.yaml]: Ah ok sounds good

File: examples/hpo/gp_mpc/config_overrides/cartpole/cartpole_stab.yaml (hunk 3 above)
Comment (Federico-PizarroBejarano) [reply examples/hpo/gp_mpc/config_overrides/cartpole/cartpole_stab.yaml:1]: Sounds good

**Review by Federico-PizarroBejarano — APPROVED** [review]
LGTM

**Review by adamhall — APPROVED** [review]
LGTM

**Pull request merged**